src/transformers/models/pop2piano/feature_extraction_pop2piano.py

```diff
@@ -0,0 +1,400 @@
+            inputs["input_features"][i] = input_feature
+
+            # compute attention_mask for input_features and add it to list
+            attention_mask = np.ones(input_features_shapes[i][:2], dtype=np.int64)
+            attention_mask = np.pad(attention_mask, padding_mask, "constant", constant_values=self.padding_value)
+            # if it is batched then we seperate each examples using zero array
+            attention_mask = np.concatenate([attention_mask, np.zeros([1, zero_array_len], dtype=np.int64)], axis=0)
+            attention_masks.append(attention_mask)
+        inputs["input_features"] = np.concatenate(inputs["input_features"], axis=0)
+        inputs["attention_mask"] = np.concatenate(attention_masks, axis=0)
+
+        attention_mask_beatsteps = []
+        for i, beatsteps in enumerate(inputs["beatsteps"]):
+            beatsteps_pad_value = max([*zip(*beatsteps_shapes)][0]) - beatsteps_shapes[i][0]
+            padding = (0, beatsteps_pad_value)
+
+            beatsteps = np.pad(beatsteps, padding, "constant", constant_values=self.padding_value)
+            inputs["beatsteps"][i] = beatsteps.reshape(1, -1)
+            # compute attention_mask for beatsteps and add it to list
+            attention_mask_beatstep = np.ones(beatsteps_shapes[i])
+            attention_mask_beatstep = np.pad(
+                attention_mask_beatstep, padding, "constant", constant_values=self.padding_value
+            ).reshape(1, -1)
+            attention_mask_beatsteps.append(attention_mask_beatstep)
+        inputs["beatsteps"] = np.concatenate(inputs["beatsteps"], axis=0)
+        inputs["attention_mask_beatsteps"] = np.concatenate(attention_mask_beatsteps, axis=0)
+
+        attention_mask_extrapolated_beatsteps = []
+        for i, extrapolated_beatstep in enumerate(inputs["extrapolated_beatstep"]):
+            extrapolated_beatstep_pad_value = (
+                max([*zip(*extrapolated_beatstep_shapes)][0]) - extrapolated_beatstep_shapes[i][0]
+            )
+            padding = (0, extrapolated_beatstep_pad_value)
+
+            extrapolated_beatstep = np.pad(
+                extrapolated_beatstep, padding, "constant", constant_values=self.padding_value
+            )
+            inputs["extrapolated_beatstep"][i] = extrapolated_beatstep.reshape(1, -1)
+            # compute attention_mask for extrapolated_beatstep and add it to list
+            attention_mask_extrapolated_beatstep = np.ones(extrapolated_beatstep_shapes[i])
+            attention_mask_extrapolated_beatstep = np.pad(
+                attention_mask_extrapolated_beatstep, padding, "constant", constant_values=self.padding_value
+            ).reshape(1, -1)
+            attention_mask_extrapolated_beatsteps.append(attention_mask_beatstep)
+        inputs["extrapolated_beatstep"] = np.concatenate(inputs["extrapolated_beatstep"], axis=0)
+        inputs["attention_mask_extrapolated_beatstep"] = np.concatenate(attention_mask_extrapolated_beatsteps, axis=0)
```

Comment:
Taking in lists as input, iterating over them in a for loop then contanenating the outputs like this is a bit of an anti pattern. 

Instead, the logic inside the for loops should be its own function, and the outputs of that function concatenated. 

Also, why is this single function padding in three for loops three independent variables? AFAICT, this logic is repeated and a more generic pad function could be called three separate times

Reply:
In the recent commit, I have broken down the large function into `self.pad` and `self._pad`, where  `self.pad` passes each different input(eg. "input_features", "beatsteps", "extrapolated_beatstep") to `self._pad`(a generic function which is called three times) each time and padding is done.